### PR TITLE
[WIP] upgrade reanimated to v3

### LINKED ISSUE #15764: Upgrade Reanimated to V3
Now that we have upgraded to RN v 0.67.5, we can upgrade to reanimated v3

### PR DESCRIPTION
**Pull request by OmarBasem**
fixes: https://github.com/status-im/status-mobile/issues/15764

### CODE REVIEW

File: nix/deps/gradle/proj.list
Comment (siddarthkay) [inline nix/deps/gradle/proj.list:1]: hmm something went wrong over here, we should revert this file to the way it was

**Comment by OmarBasem** [comment]
When running `make run-android` I get the following error:

` 
java.lang.RuntimeException: com.android.builder.sdk.InstallFailedException: Failed to install the following SDK components:
      patcher;v4 SDK Patch Applier v4
      ndk;23.1.7779620 NDK (Side by side) 23.1.7779620
`

Upgrading NDK to version `23.1.7779620` in `nix/pkgs/android-sdk/compose.nix` gives the following error:

` 
/nix/store/m4h36k1cbmhavv177mgmyg4q8vjri7b3-gomobile-unstable-2022-05-18/bin/gomobile: no usable NDK in /nix/store/n4zz2v9vadj706rbpbpa6f870ix5vl7j-androidsdk-mod-sdk: open /nix/store/n4zz2v9vadj706rbpbpa6f870ix5vl7j-androidsdk-mod-sdk/ndk/23.1.7779620/meta/platforms.json: no such file or directory, open /nix/store/n4zz2v9vadj706rbpbpa6f870ix5vl7j-androidsdk-mod-sdk/ndk-bundle/meta/platforms.json: no such file or directory
`


Also, running `make android-clean` gives the following error:

`
 Configure project :react-native-reanimated
No AAR for react-native-reanimated found. Attempting to build from source.
Android gradle plugin: 7.3.1
Gradle: 7.5
WARNING:Using flatDir should be avoided because it doesn't support any meta-data formats.
WARNING:Software Components will not be created automatically for Maven publishing from Android Gradle Plugin 8.0. To opt-in to the future behavior, set the Gradle property android.disableAutomaticComponentCreation=true in the `gradle.properties` file or use the new publishing DSL.
`

**Comment by OmarBasem** [comment]
Hi @jakubgs,
I am trying to upgrade a library (react-native-reanimated). To upgrade it, I had to upgrade gradle version and grade build tools version. I am still running into some errors on Android. Could you please [take a look](https://github.com/status-im/status-mobile/pull/15769#issuecomment-1525692226).

**Comment by jakubgs** [comment]
It doesn't seem like you changed the SDK version:
https://github.com/status-im/status-mobile/blob/fb5caa7064e824cc7de950f668d0ca321f1f6a0c/nix/pkgs/android-sdk/compose.nix#L16
If it even is available in our version of nixpkgs.

**Comment by OmarBasem** [comment]
@jakubgs I tried changing it, and then reverted it back

**Comment by OmarBasem** [comment]
I have just pushed a commit with the updated version to see if it will be any different on CI. (and it seems there is no difference)

Btw, I see the NDK version in `/nix/pkgs/android-sdk/compose.nix` do not match the ones in `/android/build.gradle`

**Comment by jakubgs** [comment]
The AAPT2 build issue is known, and was fixed in one of the RN update PRs with this: https://github.com/status-im/status-mobile/commit/d69277ed

**Comment by jakubgs** [comment]
Yes, value in `android/build.gradle` should normally reflect the version we are using from Nix.

Also, please don't add a root slash to relative paths.
It's `android/build.gradle`, not `/android/build.gradle`, pretty sure it's not at the root of your filesystem.

**Comment by OmarBasem** [comment]
> The AAPT2 build issue is known, and was fixed in one of the RN update PRs with this: [d69277ed](https://github.com/status-im/status-mobile/commit/d69277ed)

@jakubgs after doing that update to `default.nix` I am getting the following error:
`error: 1 dependencies of derivation '/nix/store/vhknay6drrb228a7hz0hpy3s4kjb0jv0-status-mobile-maven-deps.drv' failed to build
error: build of '/nix/store/1wvicqsq8p6kiqidibd36gd4pn25ajyf-status-go-0.148.3-8608aec-android.drv', '/nix/store/vhknay6drrb228a7hz0hpy3s4kjb0jv0-status-mobile-maven-deps.drv' failed
make: *** [run-android] Error 1`

**Comment by jakubgs** [comment]
No, that's not the error. That's just what is printed after the actual error.
Your builds are failing due to missing gradle deps:
```
> Could not resolve all files for configuration ':classpath'.
   > Could not find commons-codec:commons-codec:1.11.
     Searched in the following locations:
       - file:/build/android/libs/commons-codec-1.11.jar
       - file:/build/android/libs/commons-codec.jar
       - file:/build/android/app/libs/commons-codec-1.11.jar
       - file:/build/android/app/libs/commons-codec.jar
       - file:/nix/store/csq035dx5q64pxjnxdn8x4h0ap2b14s3-status-mobile-maven-deps/commons-codec/commons-codec/1.11/commons-codec-1.11.pom
     Required by:
         project : > com.android.tools.build:gradle:7.3.1 > com.android.tools.build:builder:7.3.1
```
Are you sure you updated them?

**Comment by jakubgs** [comment]
Oh, I see, running `make nix-update-gradle` is failing with:
```
* What went wrong:
The Android Gradle plugin supports only Kotlin Gradle plugin version 1.5.20 and higher.
The following dependencies do not satisfy the required version:
root project 'StatusIm' -> org.jetbrains.kotlin:kotlin-gradle-plugin:1.4.31
```
That's what the `nix/deps/gradle/add_package.sh` script was created for:
```
 > make shell TARGET=gradle
Configuring Nix shell for target 'gradle'...
Yarn modules changed, copying new version over
Copying node_modules from Nix store:
 - /nix/store/zbrx7vqcz1y3r60md3qricgv2bp7x72b-status-mobile-node-deps-1.22.0
Done in: 12.136s

[nix-shell:~/work/status-mobile]$ nix/deps/gradle/add_package.sh org.jetbrains.kotlin:kotlin-gradle-plugin:1.5.20
Changes made:
 nix/deps/gradle/deps.urls | 27 +++++++++++++++++++++++++++
 1 file changed, 27 insertions(+)

Regenerating Nix files...
Successfully added: org.jetbrains.kotlin:kotlin-gradle-plugin:1.5.20

NOTICE: Running 'make nix-update-gradle' in a new shell is recommended.
```
I've pushed a commit with updated dependencies.
For more details see: https://github.com/status-im/status-mobile/blob/develop/nix/deps/gradle/README.md#known-issues

**Comment by jakubgs** [comment]
But I see this is more complex than just that. It needs more work.

**Comment by OmarBasem** [comment]
@jakubgs Trying to run `make shell TARGET=gradle` fails with error: 


```inflating: /private/tmp/nix-build-aapt2-7.3.0-8691043.drv-0/tmp.i9ieU6xMHh/NOTICE  
patching sources
installing
error: builder for '/nix/store/wz30h0nd7vf31hjq36jqj68bdn1x8ifc-aapt2-7.3.0-8691043.drv' failed with exit code 12;
       last 8 log lines:
       > unpacking sources
       > Archive:  /nix/store/7nyznw5199fg5j1j6yawn1z17ks5kwmz-aapt2-7.3.0-8691043/aapt2-7.3.0-8691043-osx.jar
       >    creating: /private/tmp/nix-build-aapt2-7.3.0-8691043.drv-0/tmp.i9ieU6xMHh/META-INF/
       >   inflating: /private/tmp/nix-build-aapt2-7.3.0-8691043.drv-0/tmp.i9ieU6xMHh/META-INF/MANIFEST.MF
       >   inflating: /private/tmp/nix-build-aapt2-7.3.0-8691043.drv-0/tmp.i9ieU6xMHh/aapt2
       >   inflating: /private/tmp/nix-build-aapt2-7.3.0-8691043.drv-0/tmp.i9ieU6xMHh/NOTICE
       > patching sources
       > installing
       For full logs, run 'nix log /nix/store/wz30h0nd7vf31hjq36jqj68bdn1x8ifc-aapt2-7.3.0-8691043.drv'.
error: 1 dependencies of derivation '/nix/store/7fmq4lxn2sca1jz7wlc5gxa6knj6f3di-status-mobile-maven-deps.drv' failed to build
make: *** [shell] Error 1
```

**Comment by OmarBasem** [comment]
> But I see this is more complex than just that. It needs more work.

So what does it need :)

**Comment by jakubgs** [comment]
It means it needs more work. Also, please use triple backticks for large logs, otherwise the formatting is disgusting:
https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks#fenced-code-blocks

**Comment by jakubgs** [comment]
Some possibly related issues:

* https://github.com/RocketChat/Rocket.Chat.ReactNative/issues/4983
* https://github.com/software-mansion/react-native-reanimated/issues/280
* https://github.com/software-mansion/react-native-reanimated/issues/2045
* https://github.com/software-mansion/react-native-reanimated/issues/3334
* https://stackoverflow.com/questions/54858510/kotlin-compile-error-the-android-gradle-plugin-supports-only-kotlin-gradle-plu

**Comment by jakubgs** [comment]
I'm essentially switching between two errors. This one:
```
The Android Gradle plugin supports only Kotlin Gradle plugin version 1.5.20 and higher.
The following dependencies do not satisfy the required version:
root project 'StatusIm' -> org.jetbrains.kotlin:kotlin-gradle-plugin:1.3.50
```
Or this one:
```
The Android Gradle plugin supports only kotlin-android-extensions Gradle plugin version 1.6.20 and higher.
The following dependencies do not satisfy the required version:
root project 'StatusIm' -> org.jetbrains.kotlin:kotlin-gradle-plugin:1.5.20
```
@siddarthkay have you maybe dealt with this in your RN upgrade PR?

**Comment by siddarthkay** [comment]
@OmarBasem : hey quick note on going to reanimated v3, there are breaking changes for usages of v1, I had briefly tried to capture them here https://github.com/status-im/status-mobile/issues/15418/

You can also look at my branch which contains some of the changes : https://github.com/status-im/status-mobile/tree/remove-deprecated-react-native-reanimated-usage

You must get rid of all v1 usage in code first and move that to v2 and only then you should upgrade this library :)

**Comment by siddarthkay** [comment]
@OmarBasem : the change log on the official website is not helpful at all, what you must do is ensure we do not use any function that was part of v1 documentation page but absent from v2 documentation page.

v3 is backwards compatible with v2 and after you upgrade you won't have to refactor those parts to get the app running.

**Comment by jakubgs** [comment]
If we look at errors like:
```
The Android Gradle plugin supports only Kotlin Gradle plugin version 1.5.20 and higher.
The following dependencies do not satisfy the required version:
project ':react-native-hole-view' -> org.jetbrains.kotlin:kotlin-gradle-plugin:1.4.31
```
We can see that the issue is coming from `react-native-hole-view` which is indeed using `1.4.31`:
```
 > grep -C1 kotlin_version node_modules/react-native-hole-view/android/build.gradle   
buildscript {
    ext.kotlin_version = '1.4.31'

--
        classpath 'com.android.tools.build:gradle:3.5.4'
        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
        classpath "org.jetbrains.kotlin:kotlin-android-extensions:$kotlin_version"
    }
--
    implementation 'com.facebook.react:react-native:+'
    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
}
```
Which is one example of a module that might need forking to override this.

Or possibly an override using something like `allprojects`:
https://discuss.gradle.org/t/override-ext-property-in-subprojects/10276

**Comment by jakubgs** [comment]
Pushed some of my changes, but the fundamental issue is in the sub-projects that specify `1.4.31` as their Kotlin plugin version.

**Comment by jakubgs** [comment]
@siddarthkay already created a tagged release of `react-native-hole-view` here at `v2.1.3-status`:
https://github.com/status-im/react-native-hole-view/tree/v2.1.3-status

Which does make it possible to override the Kotlin Gradle Plugin version, as seen here:
```gradle
    def kotlin_version = rootProject.ext.has('RNHV_kotlinVersion') ? rootProject.ext.get('RNHV_kotlinVersion') : project.properties['RNHV_kotlinVersion']
    def gradle_version = rootProject.ext.has('RNHV_gradleVersion') ? rootProject.ext.get('RNHV_gradleVersion') : project.properties['RNHV_gradleVersion']
```
https://github.com/status-im/react-native-hole-view/blob/v2.1.3-status/android/build.gradle#L3-L4
Which in theory should work.

**Comment by jakubgs** [comment]
Using `v2.1.3-status` release does work, but now we fail on a missing Gradle dependency:
```
Regenerating Nix files...
Found 40 sub-projects...
Found 699 direct dependencies...
Found 1045 dependency URLs...
 ! Failed to fetch: https://repo.maven.apache.org/maven2/org/ow2/asm/asm-bom/9.5/asm-bom-9.5.jar
```
Or specifically, a missing JAR/AAR for the POM:
https://repo.maven.apache.org/maven2/org/ow2/asm/asm-bom/9.5/

That's an edge case that we should already be handing but we do not:
https://github.com/status-im/status-mobile/blob/6aae68ba047759f4b71e5d0a6bcd8ab764cfc5d6/nix/deps/gradle/url2json.sh#L113-L114

**Comment by jakubgs** [comment]
This is an issue that we already saw in the RN upgrade PR(https://github.com/status-im/status-mobile/pull/15203#issuecomment-1458869315):
```
* What went wrong:
Execution failed for task ':react-native-dialogs:extractReleaseAnnotations'.
> Could not resolve all files for configuration ':react-native-dialogs:detachedConfiguration1'.
   > Could not resolve com.android.tools.lint:lint-gradle:30.3.1.
     Required by:
         project :react-native-dialogs
      > No cached version of com.android.tools.lint:lint-gradle:30.3.1 available for offline mode.
      > No cached version of com.android.tools.lint:lint-gradle:30.3.1 available for offline mode.
```
For now I can add it manually, but it will disappear at next use of `make nix-update-gradle`.

**Comment by jakubgs** [comment]
Wht the fuck is this:
```
Could not evaluate onlyIf predicate for task ':react-native-reanimated:downloadBoost'.
 > Unable to download file 'boost_1_63_0.tar.gz' in offline mode.
```

**Comment by OmarBasem** [comment]
> You must get rid of all v1 usage in code first and move that to v2 and only then you should upgrade this library :)

@siddarthkay we can get rid of v1 code immediately no problem. I don't think we are using v1 code in `status-im2`

**Comment by OmarBasem** [comment]
> Could not evaluate onlyIf predicate for task ':react-native-reanimated:downloadBoost'.
>  > Unable to download file 'boost_1_63_0.tar.gz' in offline mode.

@jakubgs possibly related issue: https://github.com/software-mansion/react-native-reanimated/issues/3638

**Comment by jakubgs** [comment]
Well, that issue says to upgrade to v2, which is probably not what we want:
```
547766 newfstatat(AT_FDCWD, "/home/jakubgs/work/status-mobile/node_modules/react-native-reanimated/android/build/downloads/boost_1_63_0.tar.gz", 0x7fcdf81e6f00, 0) = -1 ENOENT (No such file or directory)
```
This can be found in the strace log from Gradle. It's looking for it in:
```
node_modules/react-native-reanimated/android/build/downloads/boost_1_63_0.tar.gz
```
Wtf is this and why it's trying to download it in some custom way? No idea.

**Comment by jakubgs** [comment]
Apparently this is only necessary when using RN older than 71:
```gradle
if (REACT_NATIVE_MINOR_VERSION < 71) {
    // You need to have following folders in this directory:
    //   - boost_1_63_0
    //   - double-conversion-1.1.6
    //   - folly-deprecate-dynamic-initializer
    //   - glog-0.3.5
    def dependenciesPath = System.getenv("REACT_NATIVE_DEPENDENCIES")

    // The Boost library is a very large download (>100MB).
    // If Boost is already present on your system, define the REACT_NATIVE_BOOST_PATH env variable
    // and the build will use that.
    def boostPath = dependenciesPath ?: System.getenv("REACT_NATIVE_BOOST_PATH")
```
https://github.com/software-mansion/react-native-reanimated/blob/3.1.0/android/build.gradle#L705-L716

Now that's some dumb shit. Do they just expect people to download those manually?

**Comment by jakubgs** [comment]
They actually download them:
```gradle
    if (REACT_NATIVE_MINOR_VERSION < 71) {
        extractAARHeaders.dependsOn(prepareThirdPartyNdkHeaders)
        extractSOFiles.dependsOn(prepareThirdPartyNdkHeaders)

        nativeBuildDependsOn(prepareThirdPartyNdkHeaders)
```
```gradle
    task prepareThirdPartyNdkHeaders(dependsOn:[
        downloadNdkBuildDependencies,
```
```gradle
    task downloadBoost(dependsOn: resolveBoost, type: Download) {
        def transformedVersion = BOOST_VERSION.replace("_", ".")
        def artifactLocalName = "boost_${BOOST_VERSION}.tar.gz"
        def srcUrl = "https://boostorg.jfrog.io/artifactory/main/release/${transformedVersion}/source/boost_${BOOST_VERSION}.tar.gz"
        if (REACT_NATIVE_MINOR_VERSION < 69) {
            srcUrl = "https://github.com/react-native-community/boost-for-react-native/releases/download/v${transformedVersion}-0/boost_${BOOST_VERSION}.tar.gz"
        }
```
But since the Nix build is pure there is no network for it to do that.

Why do they re-invent the wheel and do their own dependency management? Who thought this was a good idea?
Now we might have to write some custom Nix derivations to download this and then patch the module to make this work.

Or we could actually do the React Native upgrade...

**Comment by siddarthkay** [comment]
@jakubgs : I am resuming work on the upgrade this week, should be fun, will keep you posted.

**Comment by OmarBasem** [comment]
>  if (REACT_NATIVE_MINOR_VERSION < 71) { 

Oh I see..

**Comment by OmarBasem** [comment]
Fine we can do the RN upgrade then

**Comment by OmarBasem** [comment]
Dependency: https://github.com/status-im/status-mobile/pull/15203

**Comment by flexsurfer** [comment]
feel free to reopen when needed, or create a new one

**Pull request closed without merging**